// react-native-sortables/src/components/shared/DraggableView/ItemCell.tsx
import type { PropsWithChildren } from 'react';
import {
  type LayoutChangeEvent,
  Platform,
  StyleSheet,
  type ViewStyle
} from 'react-native';
import type {
  AnimatedStyle,
  SharedValue,
  TransformArrayItem
} from 'react-native-reanimated';
import Animated, { useAnimatedStyle } from 'react-native-reanimated';

import { HIDDEN_X_OFFSET, IS_WEB } from '../../../constants';
import type {
  AnimatedStyleProp,
  LayoutAnimation
} from '../../../integrations/reanimated';
import { useCommonValuesContext, useItemDecoration } from '../../../providers';
import AnimatedOnLayoutView from '../AnimatedOnLayoutView';

type TransformsArray = Array<TransformArrayItem>;

export type ItemCellProps = PropsWithChildren<{
  itemKey: string;
  isActive: SharedValue<boolean>;
  activationAnimationProgress: SharedValue<number>;
  baseStyle: AnimatedStyleProp;
  layoutStyleValue: SharedValue<ViewStyle>;
  hidden?: boolean;
  entering?: LayoutAnimation;
  exiting?: LayoutAnimation;
  onLayout?: (event: LayoutChangeEvent) => void;
}>;

export default function ItemCell({
  activationAnimationProgress,
  baseStyle,
  children,
  entering,
  exiting,
  hidden,
  isActive,
  itemKey,
  layoutStyleValue,
  onLayout
}: ItemCellProps) {
  const { overriddenCellDimensions } = useCommonValuesContext();

  const decorationStyleValue = useItemDecoration(
    itemKey,
    isActive,
    activationAnimationProgress
  );

  const animatedCellStyle = useAnimatedStyle(() => ({
    ...decorationStyleValue.value,
    ...layoutStyleValue.value,
    transform: [
      ...((layoutStyleValue.value.transform ?? []) as TransformsArray),
      ...((decorationStyleValue.value.transform ?? []) as TransformsArray)
    ],
    ...(!IS_WEB && overriddenCellDimensions.value)
  }));

  let innerAnimatedStyle: AnimatedStyle | undefined;
  if (IS_WEB) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    innerAnimatedStyle = useAnimatedStyle(() => overriddenCellDimensions.value);
  }

  return (
    <Animated.View style={[baseStyle, styles.decoration, animatedCellStyle]}>
      <AnimatedOnLayoutView
        entering={entering}
        exiting={exiting}
        style={[styles.inner, innerAnimatedStyle, hidden && styles.hidden]}
        onLayout={onLayout}>
        {children}
      </AnimatedOnLayoutView>
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  decoration: Platform.select<ViewStyle>({
    android: {},
    default: {},
    native: {
      shadowOffset: {
        height: 0,
        width: 0
      },
      shadowOpacity: 1,
      shadowRadius: 5
    }
  }),
  hidden: {
    // We change the x position to hide items when teleported (we can't use
    // non-layout props like opacity as they are sometimes not updated via
    // Reanimated on the Old Architecture; we also can't use any props that
    // affect item dimensions, etc., so the safest way is to put the item
    // far away from the viewport to hide it)
    left: HIDDEN_X_OFFSET
  },
  inner: Platform.select<ViewStyle>({
    default: {},
    web: { flex: 1 }
  })
});
