// reanimated-tab-view/lib/module/components/Scrollable/RTVScrollView.js
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import React, { forwardRef, useImperativeHandle, useMemo } from 'react';
import { Gesture, GestureDetector } from 'react-native-gesture-handler';
import Animated, { runOnJS, useAnimatedRef, useAnimatedScrollHandler, useAnimatedStyle, useSharedValue } from 'react-native-reanimated';
import { StyleSheet } from 'react-native';
import { useScrollableContext } from '../../providers/Scrollable';
import { useInternalContext } from '../../providers/Internal';
import { useScrollHandlers } from '../../hooks/scrollable/useScrollHandlers';
import { useSyncScrollWithPanTranslation } from '../../hooks/scrollable/useSyncScrollWithPanTranslation';
import { SHOULD_RENDER_ABSOLUTE_HEADER } from '../../constants/scrollable';
export const RTVScrollView = /*#__PURE__*/React.memo( /*#__PURE__*/forwardRef((props, ref) => {
  //#region props
  const {
    children,
    onScroll: _onScroll,
    onScrollEndDrag: _onScrollEndDrag,
    onScrollBeginDrag: _onScrollBeginDrag,
    onMomentumScrollEnd: _onMomentumScrollEnd,
    onMomentumScrollBegin: _onMomentumScrollBegin,
    ...rest
  } = props;
  //#endregion

  //#region context
  const {
    animatedTranslateYSV
  } = useScrollableContext();
  const {
    tabViewHeaderLayout,
    tabBarLayout,
    tabViewCarouselLayout
  } = useInternalContext();

  //#endregion

  //#region variables
  const scrollRef = useAnimatedRef();
  const scrollGesture = useMemo(() => Gesture.Native().shouldCancelWhenOutside(false).disallowInterruption(true), []);
  const scrollYSV = useSharedValue(0);
  const {
    onBeginDrag,
    onScroll
  } = useScrollHandlers(scrollYSV);
  //#endregion

  //#region styles
  const animatedContentContainerStyle = useAnimatedStyle(() => {
    return {
      transform: [{
        translateY: animatedTranslateYSV.value
      }]
    };
  }, [animatedTranslateYSV]);
  const translatingContentContainerStyle = useMemo(() => {
    return {
      ...animatedContentContainerStyle,
      paddingBottom: tabViewHeaderLayout.height,
      minHeight: tabViewCarouselLayout.height + tabViewHeaderLayout.height
    };
  }, [animatedContentContainerStyle, tabViewCarouselLayout.height, tabViewHeaderLayout.height]);
  const nonTranslatingContentContainerStyle = useMemo(() => {
    return {
      paddingTop: tabBarLayout.height + tabViewHeaderLayout.height,
      minHeight: tabViewCarouselLayout.height + tabViewHeaderLayout.height
    };
  }, [tabBarLayout.height, tabViewCarouselLayout.height, tabViewHeaderLayout.height]);
  //#endregion

  //#region callbacks
  const handleScroll = useAnimatedScrollHandler({
    onScroll: event => {
      onScroll(event);
      if (_onScroll) {
        runOnJS(_onScroll)({
          nativeEvent: event
        });
      }
    },
    onBeginDrag: event => {
      onBeginDrag();
      if (_onScrollBeginDrag) {
        runOnJS(_onScrollBeginDrag)({
          nativeEvent: event
        });
      }
    },
    onEndDrag: event => {
      if (_onScrollEndDrag) {
        runOnJS(_onScrollEndDrag)({
          nativeEvent: event
        });
      }
    },
    onMomentumEnd: event => {
      if (_onMomentumScrollEnd) {
        runOnJS(_onMomentumScrollEnd)({
          nativeEvent: event
        });
      }
    },
    onMomentumBegin: event => {
      if (_onMomentumScrollBegin) {
        runOnJS(_onMomentumScrollBegin)({
          nativeEvent: event
        });
      }
    }
  });
  //#endregion

  //#region hooks
  useImperativeHandle(ref, () => scrollRef.current);
  useSyncScrollWithPanTranslation(scrollRef, scrollYSV);
  //#endregion

  //#region render
  return /*#__PURE__*/React.createElement(GestureDetector, {
    gesture: scrollGesture
  }, /*#__PURE__*/React.createElement(Animated.ScrollView, _extends({
    ref: scrollRef
  }, rest, {
    onScroll: handleScroll,
    scrollEventThrottle: 16
  }), SHOULD_RENDER_ABSOLUTE_HEADER ? /*#__PURE__*/React.createElement(Animated.View, {
    style: [styles.contentContainer, nonTranslatingContentContainerStyle]
  }, children) : /*#__PURE__*/React.createElement(Animated.View, {
    style: [styles.contentContainer, translatingContentContainerStyle]
  }, children)));
  //#endregion
}));
const styles = StyleSheet.create({
  contentContainer: {
    flex: 1
  }
});
