// @codeherence/react-native-header/lib/module/components/containers/FadingView.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import React, { forwardRef } from 'react';
import { StyleSheet } from 'react-native';
import Animated, { useAnimatedProps, useAnimatedStyle } from 'react-native-reanimated';
const FadingView = /*#__PURE__*/forwardRef(({
  children,
  style,
  opacity,
  animatedProps = {},
  opacityThresholdToEnablePointerEvents = 1,
  ...rest
}, ref) => {
  const _animatedProps = useAnimatedProps(() => {
    const _pointerEvents = opacity.value >= opacityThresholdToEnablePointerEvents ? 'auto' : 'none';
    return {
      pointerEvents: _pointerEvents
    };
  }, [opacityThresholdToEnablePointerEvents]);
  const fadeStyle = useAnimatedStyle(() => ({
    opacity: opacity.value
  }));
  return /*#__PURE__*/React.createElement(Animated.View, _extends({
    ref: ref,
    style: [styles.container, style, fadeStyle],
    animatedProps: {
      ..._animatedProps,
      ...animatedProps
    }
  }, rest), children);
});
export default FadingView;
const styles = StyleSheet.create({
  container: {
    opacity: 0
  }
});
