// @codeherence/react-native-header/lib/module/components/containers/SectionList.js
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
import React from 'react';
import { View, SectionList, StyleSheet } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import Animated, { useAnimatedRef } from 'react-native-reanimated';
import FadingView from '../containers/FadingView';
import { useScrollContainerLogic } from './useScrollContainerLogic';
const AnimatedSectionList = Animated.createAnimatedComponent(SectionList);
const SectionListWithHeaders = ({
  largeHeaderShown,
  containerStyle,
  LargeHeaderSubtitleComponent,
  LargeHeaderComponent,
  largeHeaderContainerStyle,
  HeaderComponent,
  onLargeHeaderLayout,
  onScrollBeginDrag,
  onScrollEndDrag,
  onScrollWorklet,
  onMomentumScrollBegin,
  onMomentumScrollEnd,
  ignoreLeftSafeArea,
  ignoreRightSafeArea,
  disableAutoFixScroll = false,
  // We use this to ensure that the onScroll property isn't accidentally used.
  // @ts-ignore
  onScroll: _unusedOnScroll,
  absoluteHeader = false,
  initialAbsoluteHeaderHeight = 0,
  contentContainerStyle,
  automaticallyAdjustsScrollIndicatorInsets,
  headerFadeInThreshold = 1,
  disableLargeHeaderFadeAnim = false,
  scrollIndicatorInsets = {},
  inverted,
  ...rest
}) => {
  if (_unusedOnScroll) {
    throw new Error("The 'onScroll' property is not supported. Please use onScrollWorklet to track the scroll container's state.");
  }
  const insets = useSafeAreaInsets();
  const scrollRef = useAnimatedRef();
  const {
    scrollY,
    showNavBar,
    largeHeaderHeight,
    largeHeaderOpacity,
    scrollHandler,
    debouncedFixScroll,
    onAbsoluteHeaderLayout,
    scrollViewAdjustments
  } = useScrollContainerLogic({
    scrollRef,
    largeHeaderShown,
    disableAutoFixScroll,
    largeHeaderExists: !!LargeHeaderComponent,
    absoluteHeader,
    initialAbsoluteHeaderHeight,
    headerFadeInThreshold,
    inverted: !!inverted,
    onScrollWorklet
  });
  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, containerStyle, !ignoreLeftSafeArea && {
      paddingLeft: insets.left
    }, !ignoreRightSafeArea && {
      paddingRight: insets.right
    }]
  }, !absoluteHeader && HeaderComponent({
    showNavBar,
    scrollY
  }), /*#__PURE__*/React.createElement(AnimatedSectionList
  // ref={scrollRef} // TODO: Fix this since createAnimatedComponent is not working with SectionList
  , _extends({
    scrollEventThrottle: 16,
    overScrollMode: "auto",
    onScroll: scrollHandler,
    automaticallyAdjustContentInsets: false,
    onScrollBeginDrag: e => {
      debouncedFixScroll.cancel();
      if (onScrollBeginDrag) onScrollBeginDrag(e);
    },
    onScrollEndDrag: e => {
      debouncedFixScroll();
      if (onScrollEndDrag) onScrollEndDrag(e);
    },
    onMomentumScrollBegin: e => {
      debouncedFixScroll.cancel();
      if (onMomentumScrollBegin) onMomentumScrollBegin(e);
    },
    onMomentumScrollEnd: e => {
      debouncedFixScroll();
      if (onMomentumScrollEnd) onMomentumScrollEnd(e);
    },
    contentContainerStyle: [scrollViewAdjustments.contentContainerStyle,
    // @ts-ignore
    // Unfortunately there are issues with Reanimated typings, so will ignore for now.
    contentContainerStyle],
    automaticallyAdjustsScrollIndicatorInsets: automaticallyAdjustsScrollIndicatorInsets !== undefined ? automaticallyAdjustsScrollIndicatorInsets : !absoluteHeader,
    scrollIndicatorInsets: {
      ...scrollViewAdjustments.scrollIndicatorInsets,
      ...scrollIndicatorInsets
    },
    ListHeaderComponent: /*#__PURE__*/React.createElement(React.Fragment, null, LargeHeaderComponent && /*#__PURE__*/React.createElement(View, {
      onLayout: e => {
        largeHeaderHeight.value = e.nativeEvent.layout.height;
        if (onLargeHeaderLayout) onLargeHeaderLayout(e.nativeEvent.layout);
      }
    }, !disableLargeHeaderFadeAnim ? /*#__PURE__*/React.createElement(FadingView, {
      opacity: largeHeaderOpacity,
      style: largeHeaderContainerStyle
    }, LargeHeaderComponent({
      scrollY,
      showNavBar
    })) : /*#__PURE__*/React.createElement(View, {
      style: largeHeaderContainerStyle
    }, LargeHeaderComponent({
      scrollY,
      showNavBar
    }))), LargeHeaderSubtitleComponent && LargeHeaderSubtitleComponent({
      showNavBar,
      scrollY
    })),
    inverted: inverted
  }, rest)), absoluteHeader && /*#__PURE__*/React.createElement(View, {
    style: styles.absoluteHeader,
    onLayout: onAbsoluteHeaderLayout
  }, HeaderComponent({
    showNavBar,
    scrollY
  })));
};
export default SectionListWithHeaders;
const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  absoluteHeader: {
    position: 'absolute',
    top: 0,
    right: 0,
    left: 0
  }
});
