// @codeherence/react-native-header/lib/commonjs/components/containers/FlashList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeSafeAreaContext = require("react-native-safe-area-context");
var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));
var _flashList = require("@shopify/flash-list");
var _FadingView = _interopRequireDefault(require("./FadingView"));
var _useScrollContainerLogic = require("./useScrollContainerLogic");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const AnimatedFlashList = _reactNativeReanimated.default.createAnimatedComponent(_flashList.FlashList);
const FlashListWithHeadersInputComp = ({
  largeHeaderShown,
  containerStyle,
  LargeHeaderSubtitleComponent,
  LargeHeaderComponent,
  largeHeaderContainerStyle,
  HeaderComponent,
  onLargeHeaderLayout,
  onScrollBeginDrag,
  onScrollEndDrag,
  onScrollWorklet,
  onMomentumScrollBegin,
  onMomentumScrollEnd,
  ignoreLeftSafeArea,
  ignoreRightSafeArea,
  disableAutoFixScroll = false,
  // We use this to ensure that the onScroll property isn't accidentally used.
  // @ts-ignore
  onScroll: _unusedOnScroll,
  absoluteHeader = false,
  initialAbsoluteHeaderHeight = 0,
  contentContainerStyle = {},
  automaticallyAdjustsScrollIndicatorInsets,
  headerFadeInThreshold = 1,
  disableLargeHeaderFadeAnim = false,
  scrollIndicatorInsets = {},
  ...rest
}, ref) => {
  if (_unusedOnScroll) {
    throw new Error("The 'onScroll' property is not supported. Please use onScrollWorklet to track the scroll container's state.");
  }
  const insets = (0, _reactNativeSafeAreaContext.useSafeAreaInsets)();
  const scrollRef = (0, _reactNativeReanimated.useAnimatedRef)();
  (0, _react.useImperativeHandle)(ref, () => scrollRef.current);
  const {
    maintainVisibleContentPosition
  } = rest;
  const inverted = (0, _react.useMemo)(() => {
    if (maintainVisibleContentPosition === undefined) return false;
    if ((0, _reactNativeReanimated.isSharedValue)(maintainVisibleContentPosition)) return false;
    return (maintainVisibleContentPosition === null || maintainVisibleContentPosition === void 0 ? void 0 : maintainVisibleContentPosition.startRenderingFromBottom) ?? false;
  }, [maintainVisibleContentPosition]);
  const {
    scrollY,
    showNavBar,
    largeHeaderHeight,
    largeHeaderOpacity,
    scrollHandler,
    debouncedFixScroll,
    onAbsoluteHeaderLayout,
    scrollViewAdjustments
  } = (0, _useScrollContainerLogic.useScrollContainerLogic)({
    scrollRef,
    largeHeaderShown,
    disableAutoFixScroll,
    largeHeaderExists: !!LargeHeaderComponent,
    absoluteHeader,
    initialAbsoluteHeaderHeight,
    headerFadeInThreshold,
    inverted,
    onScrollWorklet,
    isFlashList: true
  });
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, containerStyle, !ignoreLeftSafeArea && {
      paddingLeft: insets.left
    }, !ignoreRightSafeArea && {
      paddingRight: insets.right
    }]
  }, !absoluteHeader && HeaderComponent({
    showNavBar,
    scrollY
  }), /*#__PURE__*/_react.default.createElement(AnimatedFlashList, _extends({
    ref: scrollRef,
    scrollEventThrottle: 16,
    overScrollMode: "auto",
    onScroll: scrollHandler,
    automaticallyAdjustContentInsets: false,
    onScrollBeginDrag: e => {
      debouncedFixScroll.cancel();
      if (onScrollBeginDrag) onScrollBeginDrag(e);
    },
    onScrollEndDrag: e => {
      debouncedFixScroll();
      if (onScrollEndDrag) onScrollEndDrag(e);
    },
    onMomentumScrollBegin: e => {
      debouncedFixScroll.cancel();
      if (onMomentumScrollBegin) onMomentumScrollBegin(e);
    },
    onMomentumScrollEnd: e => {
      debouncedFixScroll();
      if (onMomentumScrollEnd) onMomentumScrollEnd(e);
    },
    contentContainerStyle: {
      // The reason why we do this is because FlashList does not support an array of
      // styles (will throw a warning when you supply one).
      ...scrollViewAdjustments.contentContainerStyle,
      ...contentContainerStyle
    },
    automaticallyAdjustsScrollIndicatorInsets: automaticallyAdjustsScrollIndicatorInsets !== undefined ? automaticallyAdjustsScrollIndicatorInsets : !absoluteHeader,
    scrollIndicatorInsets: {
      ...scrollViewAdjustments.scrollIndicatorInsets,
      ...scrollIndicatorInsets
    },
    ListHeaderComponent: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, LargeHeaderComponent && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      onLayout: e => {
        largeHeaderHeight.value = e.nativeEvent.layout.height;
        if (onLargeHeaderLayout) onLargeHeaderLayout(e.nativeEvent.layout);
      }
    }, !disableLargeHeaderFadeAnim ? /*#__PURE__*/_react.default.createElement(_FadingView.default, {
      opacity: largeHeaderOpacity,
      style: largeHeaderContainerStyle
    }, LargeHeaderComponent({
      scrollY,
      showNavBar
    })) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
      style: largeHeaderContainerStyle
    }, LargeHeaderComponent({
      scrollY,
      showNavBar
    }))), LargeHeaderSubtitleComponent && LargeHeaderSubtitleComponent({
      showNavBar,
      scrollY
    }))
  }, rest)), absoluteHeader && /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.absoluteHeader,
    onLayout: onAbsoluteHeaderLayout
  }, HeaderComponent({
    showNavBar,
    scrollY
  })));
};
const FlashListWithHeaders = /*#__PURE__*/_react.default.forwardRef(FlashListWithHeadersInputComp);
var _default = exports.default = FlashListWithHeaders;
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1
  },
  absoluteHeader: {
    position: 'absolute',
    top: 0,
    right: 0,
    left: 0
  }
});
