// @codeherence/react-native-header/lib/commonjs/components/headers/Header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _reactNativeSafeAreaContext = require("react-native-safe-area-context");
var _containers = require("../containers");
var _line = require("../line");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const MIN_CENTER_WIDTH_PRC = 0.4;
const Header = ({
  showNavBar,
  headerStyle,
  headerLeft = null,
  headerLeftStyle,
  headerLeftFadesIn,
  headerCenter = null,
  headerCenterStyle,
  headerCenterFadesIn = true,
  headerRight = null,
  headerRightStyle,
  headerRightFadesIn,
  noBottomBorder = false,
  ignoreTopSafeArea = false,
  borderColor,
  initialBorderColor,
  borderWidth,
  SurfaceComponent
}) => {
  const {
    top
  } = (0, _reactNativeSafeAreaContext.useSafeAreaInsets)();
  const dimensions = (0, _reactNative.useWindowDimensions)();
  const headerCenterExists = !!headerCenter;
  const {
    centerWidth,
    minSideHeaderWidth
  } = (0, _react.useMemo)(() => {
    const _centerWidth = headerCenterExists ? MIN_CENTER_WIDTH_PRC * dimensions.width : 0;
    const _minSideHeaderWidth = (dimensions.width - _centerWidth) / 2;
    return {
      centerWidth: _centerWidth,
      minSideHeaderWidth: _minSideHeaderWidth
    };
  }, [headerCenterExists, dimensions]);
  const noHeaderLeftRight = !headerLeft && !headerRight;
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, null, SurfaceComponent && SurfaceComponent({
    showNavBar
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, !ignoreTopSafeArea && {
      paddingTop: top
    }, headerStyle]
  }, headerLeftFadesIn ? /*#__PURE__*/_react.default.createElement(_containers.FadingView, {
    opacity: showNavBar,
    style: [styles.leftContainer, noHeaderLeftRight && styles.noFlex, {
      width: minSideHeaderWidth
    }, headerLeftStyle]
  }, headerLeft) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.leftContainer, noHeaderLeftRight && styles.noFlex, {
      width: minSideHeaderWidth
    }, headerLeftStyle]
  }, headerLeft), headerCenter && (headerCenterFadesIn ? /*#__PURE__*/_react.default.createElement(_containers.FadingView, {
    opacity: showNavBar,
    style: [styles.centerContainer, {
      minWidth: centerWidth
    }, headerCenterStyle]
  }, headerCenter) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.centerContainer, {
      width: centerWidth
    }, headerCenterStyle]
  }, headerCenter)), headerRightFadesIn ? /*#__PURE__*/_react.default.createElement(_containers.FadingView, {
    opacity: showNavBar,
    style: [styles.rightContainer, noHeaderLeftRight && styles.noFlex, {
      width: minSideHeaderWidth
    }, headerRightStyle]
  }, headerRight) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.rightContainer, noHeaderLeftRight && styles.noFlex, {
      width: minSideHeaderWidth
    }, headerRightStyle]
  }, headerRight)), !noBottomBorder && /*#__PURE__*/_react.default.createElement(_line.HeaderBottomBorder, {
    opacity: showNavBar,
    borderColor: borderColor,
    initialBorderColor: initialBorderColor,
    borderWidth: borderWidth
  }));
};
var _default = exports.default = Header;
const styles = _reactNative.StyleSheet.create({
  container: {
    flexDirection: 'row',
    width: '100%',
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  leftContainer: {
    flexDirection: 'row',
    paddingHorizontal: 8,
    paddingVertical: 4,
    justifyContent: 'flex-start',
    alignItems: 'center',
    overflow: 'hidden'
  },
  centerContainer: {
    flex: 1,
    flexDirection: 'row',
    paddingHorizontal: 6,
    alignItems: 'center',
    justifyContent: 'center'
  },
  rightContainer: {
    flexDirection: 'row-reverse',
    paddingVertical: 4,
    paddingHorizontal: 8,
    alignItems: 'center',
    justifyContent: 'flex-start',
    overflow: 'hidden'
  },
  noFlex: {
    display: 'none'
  }
});
