// @vibecodeapp/sdk/dist/dev/StatusBar.web.js
/**
 * Web version of StatusBar - posts color to parent iframe
 *
 * On web, StatusBar doesn't render anything visible.
 * It just posts the status bar color to the parent iframe.
 */
import { useEffect } from 'react';
function postColorToParent(color) {
    try {
        if (typeof window !== 'undefined' &&
            window.parent &&
            window.parent !== window) {
            window.parent.postMessage({ type: 'vibecode:statusbar:color', color, timestamp: Date.now() }, '*');
        }
    }
    catch {
        // Silently fail
    }
}
function styleToColor(style = 'auto') {
    if (style === 'dark') {
        return '#000000';
    }
    return '#FFFFFF';
}
export function StatusBar({ style = 'auto' }) {
    useEffect(() => {
        postColorToParent(styleToColor(style));
    }, [style]);
    return null;
}
// No-op functions for web
export const setStatusBarStyle = (style, _animated) => {
    postColorToParent(styleToColor(style));
};
export const setStatusBarHidden = (_hidden, _animation) => { };
export const setStatusBarBackgroundColor = (_backgroundColor, _animated) => { };
export const setStatusBarNetworkActivityIndicatorVisible = (_visible) => { };
export const setStatusBarTranslucent = (_translucent) => { };
