// @nandorojo/galeria/build/GaleriaView.ios.jsx
import { requireNativeView } from 'expo';
import { useContext } from 'react';
import { GaleriaContext } from './context';
import { Image } from 'react-native';
const NativeImage = requireNativeView('Galeria');
const array = [];
const noop = () => { };
const Galeria = Object.assign(function Galeria({ children, closeIconName, urls, theme = 'dark', ids, }) {
    return (<GaleriaContext.Provider value={{
            closeIconName,
            urls,
            theme,
            initialIndex: 0,
            open: false,
            src: '',
            setOpen: noop,
            ids,
        }}>
        {children}
      </GaleriaContext.Provider>);
}, {
    Image(props) {
        const { theme, urls, initialIndex, closeIconName } = useContext(GaleriaContext);
        return (<NativeImage closeIconName={closeIconName} theme={theme} urls={urls?.map((url) => {
                if (typeof url === 'string') {
                    return url;
                }
                return Image.resolveAssetSource(url).uri;
            })} index={initialIndex} {...props}/>);
    },
    Popup: (() => null),
});
export default Galeria;
