// react-native-sortables/dist/module/types/state.js
"use strict";

export let DragActivationState = /*#__PURE__*/function (DragActivationState) {
  DragActivationState["INACTIVE"] = "INACTIVE";
  DragActivationState["TOUCHED"] = "TOUCHED";
  DragActivationState["ACTIVE"] = "ACTIVE";
  return DragActivationState;
}({});
export let LayerState = /*#__PURE__*/function (LayerState) {
  LayerState[LayerState["IDLE"] = 0] = "IDLE";
  LayerState[LayerState["INTERMEDIATE"] = 1] = "INTERMEDIATE";
  LayerState[LayerState["FOCUSED"] = 2] = "FOCUSED";
  return LayerState;
}({});
export let ItemPortalState = /*#__PURE__*/function (ItemPortalState) {
  /**
   * Initial state when there is no portal or the item is rendered
   * in its original position (not teleported).
   */
  ItemPortalState[ItemPortalState["IDLE"] = 0] = "IDLE";
  /**
   * Intermediate state when item teleportation has been scheduled
   * but not completed yet. Represents the transition from IDLE to
   * TELEPORTED state.
   */
  ItemPortalState[ItemPortalState["TELEPORTING"] = 1] = "TELEPORTING";
  /**
   * State when the item is fully rendered within the portal outlet
   * at its destination position.
   */
  ItemPortalState[ItemPortalState["TELEPORTED"] = 2] = "TELEPORTED";
  /**
   * Intermediate state when the item is being removed from the portal
   * outlet but the removal is not complete yet. Represents the
   * transition from TELEPORTED back to IDLE state.
   */
  ItemPortalState[ItemPortalState["EXITING"] = 3] = "EXITING";
  return ItemPortalState;
}({});
