// react-native-sortables/dist/module/utils/equality.js
"use strict";
'worklet';

export const lt = (a, b) => {
  'worklet';

  return a < b;
};
export const gt = (a, b) => {
  'worklet';

  return a > b;
};
export const areArraysDifferent = (arr1, arr2, areEqual = (a, b) => a === b) => {
  'worklet';

  return arr1.length !== arr2.length || arr1.some((item, index) => !areEqual(item, arr2[index]));
};
export const areValuesDifferent = (dim1, dim2, eps) => {
  'worklet';

  if (dim1 === undefined) {
    return dim2 !== undefined;
  }
  if (dim2 === undefined) {
    return true;
  }
  if (eps) {
    return Math.abs(dim1 - dim2) > eps;
  }
  return dim1 !== dim2;
};
export const areVectorsDifferent = (vec1, vec2, eps) => {
  'worklet';

  return areValuesDifferent(vec1.x, vec2.x, eps) || areValuesDifferent(vec1.y, vec2.y, eps);
};
export const haveEqualPropValues = (obj1, obj2) => {
  'worklet';

  if (!obj1 || !obj2) {
    return false;
  }
  const obj1Keys = Object.keys(obj1);
  const obj2Keys = Object.keys(obj2);
  return obj1Keys.length === obj2Keys.length && obj1Keys.every(key => obj1[key] === obj2[key]);
};
