// @vibecodeapp/sdk/dist/dev/error-boundry/ErrorBoundary.web.js
/**
 * Web version of ErrorBoundary - posts errors to parent iframe
 *
 * Catches React errors and global errors, sends them to parent via postMessage.
 * No fallback UI - just reports errors and continues rendering.
 * Hides Expo's error overlay by removing it when it appears.
 */
// Auto install fetch proxy for web
// Must be first import
// Should move to VibeDevWrapper
import '../../polyfills/fetch.web';
import React from 'react';
// Hide Expo's error overlay by removing it when it appears
if (typeof window !== 'undefined' && typeof MutationObserver !== 'undefined') {
    const observer = new MutationObserver((mutations) => {
        for (const mutation of mutations) {
            mutation.addedNodes.forEach((node) => {
                if (node instanceof HTMLElement && node.id === 'error-overlay') {
                    node.remove();
                }
            });
        }
    });
    observer.observe(document.body, { childList: true });
}
function sendErrorToParent(error, errorInfo) {
    try {
        if (typeof window !== 'undefined' &&
            window.parent &&
            window.parent !== window) {
            const err = error;
            window.parent.postMessage({
                type: 'vibecode:error',
                error: {
                    message: err?.message || String(error) || 'Unknown error',
                    stack: err?.stack,
                    componentStack: errorInfo?.componentStack,
                    timestamp: Date.now(),
                },
            }, '*');
        }
    }
    catch {
        // Silently fail
    }
}
// Global error listeners
if (typeof window !== 'undefined') {
    window.addEventListener('error', (event) => {
        event.preventDefault();
        sendErrorToParent(event.error ?? { message: event.message });
    }, true);
    window.addEventListener('unhandledrejection', (event) => {
        event.preventDefault();
        sendErrorToParent(event.reason);
    }, true);
    // Override console.error to intercept errors before they trigger the overlay
    const originalConsoleError = console.error;
    console.error = (...args) => {
        sendErrorToParent(args.map((a) => String(a)).join(' '));
        originalConsoleError.apply(console, args);
    };
}
export class VibeErrorBoundary extends React.Component {
    constructor() {
        super(...arguments);
        this.state = { hasError: false };
    }
    static getDerivedStateFromError() {
        return { hasError: true };
    }
    componentDidCatch(error, errorInfo) {
        sendErrorToParent(error, errorInfo);
    }
    render() {
        if (this.state.hasError) {
            // Return null to prevent infinite error loops
            // The error has already been sent to the parent iframe
            return null;
        }
        return this.props.children;
    }
}
